
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Overview &#8212; Parent-map 1.1.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Parent-map documentation" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>Parent-map analyzes protein or DNA sequences which are derived from one or multiple parent sequences, and shows parental contributions as well as differences from relevant parents.</p>
<p>Originally developed to analyze capsid protein sequences obtained by directed evolution, Parent-map can be used in any case where variant sequences are to be compared to parent sequences from which they are derived. Parent-map detects sequence shuffling as well as substitutions, insertions and deletions.</p>
<p>Parent-map takes two FASTA files as input and generates four text files and one html file as output.</p>
<p>The first input file contains the variant sequences to be analyzed, in either single- or multi-line FASTA format. Sequences are read into memory one at a time, so there is no limit to their number. They don’t need to be related, and can be derived from any combination of parental libraries.</p>
<p>The other input file contains the parental sequences. FASTA format is only required if more than one parent is present. All parental sequences are loaded into memory together, as their number is limited.</p>
<p>The output files have the following endings in their name: -stats.txt (statistics), -par.txt (parental contributions), -par.html (parental contributions, colorized html file), -def.txt (sequence definitions) and -aln.txt (alignments).</p>
<p>The -stats.txt file is a simple table showing an overview of the variant sequences main features. Sequences are sorted by the number of contributing parents, then by the parent name.</p>
<p>The -par.txt file is an alignment of each variant against all its contributing parents. This file is particularly useful for variants derived from shuffled libraries, as it maps the most likely parental contributions. Contributing parental fragments are displayed using an identity character (which can be customized). Obviously, in the case of multiple parents, the combination of fragments shown is only one among a very large number of possible combinations. However, the solution shown has the largest fragment lengths, the smallest number of fragments and the smallest number of parents.</p>
<p>The -par.html file is a colorized version of the -par.txt file, with each parent represented with a different color.</p>
<p>The -def.txt file gives a formal definition of each variant sequence in relation to its parent(s), in a table format. It’s a blueprint allowing to reconstruct the variant sequences from a minimal set of instructions.</p>
<p>The -aln file aligns variants under their main parent. As opposed to the -par.txt file where there is an alignment per variant, here there is an alignment per parent. First, alignments for single parent sequences are shown, then alignments for multi-parent sequences if they exist, against their main parent. For legibility, only regions in which at least one of the variants differ from the parent are shown. Each region is numbered according to its position in the parental sequence.</p>
<p>Parent-map is a platform-independent Python 3 script available as a bioconda package, as well as a Windows executable. It can be run in command-line mode inside a terminal or in a graphic user interface. Starting it without arguments will launch the graphic interface, while adding any argument will run it as a command-line program. The Windows version can also do both, with the only difference that it will not display anything in command-line mode.</p>
<dl class="simple">
<dt>Source code:</dt><dd><p><a class="reference external" href="https://github.com/damienmarsic/Parent-map">https://github.com/damienmarsic/Parent-map</a></p>
</dd>
<dt>Python package:</dt><dd><p><a class="reference external" href="https://pypi.org/project/parent-map/">https://pypi.org/project/parent-map/</a></p>
</dd>
<dt>Bioconda recipe and package:</dt><dd><p><a class="reference external" href="http://bioconda.github.io/recipes/parent-map/README.html">http://bioconda.github.io/recipes/parent-map/README.html</a></p>
</dd>
<dt>Windows executable:</dt><dd><p><a class="reference external" href="https://sourceforge.net/projects/parent-map/">https://sourceforge.net/projects/parent-map/</a></p>
</dd>
</dl>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>The recommended method is to install parent-map as a bioconda package under a Python 3 conda environment. If you don’t have such environment, first install Miniconda3 from here (exists for Linux, MacOSX and Windows):
<a class="reference external" href="https://docs.conda.io/en/latest/miniconda.html">https://docs.conda.io/en/latest/miniconda.html</a></p>
<p>To install the parent-map bioconda package, at the conda prompt type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span>
</pre></div>
</div>
<p>Windows users who prefer not to use a conda environment can install the Windows executable version of parent-map from here:
<a class="reference external" href="https://sourceforge.net/projects/parent-map/">https://sourceforge.net/projects/parent-map/</a></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>To run parent-map from a Python 3 conda environment, type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span>
</pre></div>
</div>
<p>This will launch the graphic interface (same as if launching the Windows version).</p>
<p>To run in command-line mode, add arguments. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span> <span class="o">-</span><span class="n">v</span>
<span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>In the first example, the version will be displayed. The second example will show a help message.</p>
<p>Typical usage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span> <span class="n">variant_file</span> <span class="n">parent_file</span> <span class="n">optional_arguments</span>
</pre></div>
</div>
<p>where variant_file is the name of the file containing the variant sequences, parent_file is the name of the file containing the parental sequences, and optional_arguments can be any combination of the optional arguments described in detail below.</p>
<p>Real-life example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span> <span class="n">example_variants</span><span class="o">.</span><span class="n">fasta</span> <span class="n">example_parents</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">o</span> <span class="n">test</span>
</pre></div>
</div>
<p>This will result in the following output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Sequence</span> <span class="n">name</span> <span class="s1">&#39;AAV2-sept-Y-F&#39;</span> <span class="n">replaced</span> <span class="k">with</span> <span class="s1">&#39;S16&#39;</span>
<span class="n">Parental</span> <span class="n">maps</span> <span class="n">saved</span> <span class="n">into</span> <span class="n">file</span><span class="p">:</span> <span class="n">test</span><span class="o">-</span><span class="n">par</span><span class="o">.</span><span class="n">txt</span>
<span class="n">Stats</span> <span class="n">saved</span> <span class="n">into</span> <span class="n">file</span><span class="p">:</span> <span class="n">test</span><span class="o">-</span><span class="n">stats</span><span class="o">.</span><span class="n">txt</span>
<span class="n">Sequence</span> <span class="n">definitions</span> <span class="n">saved</span> <span class="n">into</span> <span class="n">file</span><span class="p">:</span> <span class="n">test</span><span class="o">-</span><span class="n">def</span><span class="o">.</span><span class="n">txt</span>
<span class="n">Alignments</span> <span class="n">saved</span> <span class="n">into</span> <span class="n">file</span><span class="p">:</span> <span class="n">test</span><span class="o">-</span><span class="n">aln</span><span class="o">.</span><span class="n">txt</span>
<span class="n">Colorized</span> <span class="n">version</span> <span class="n">of</span> <span class="n">parental</span> <span class="nb">map</span> <span class="n">file</span> <span class="n">saved</span> <span class="n">into</span> <span class="n">file</span><span class="p">:</span> <span class="n">test</span><span class="o">-</span><span class="n">par</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
<p>Note that the prefix ‘test’ entered with the -o (output) argument became the prefix of the 4 output files. Also note that the sequence name change (first line) can be avoided by using the -s argument (maximum name size) followed by the value 13 (the sequence name ‘AAV2-sept-Y-F’ is 13 character long, while the default value for -s is 12).</p>
<p>Parent-map can also bypass the sequence analysis and just colorize an existing -par.txt file into a -par.html file. This is useful when the user is not satisfied with the default colors and wants to try different color schemes. In this case, the -C argument is used followed by the name of the -par.txt file. Also, the -p argument must be used followed by a list of parent color pairs. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">parent</span><span class="o">-</span><span class="nb">map</span> <span class="o">-</span><span class="n">d</span> <span class="o">-</span><span class="n">C</span> <span class="n">test</span><span class="o">-</span><span class="n">par</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">p</span> <span class="s1">&#39;AAV1 lavender, AAV2 cyan, AAV3B gold, AAV5 crimson, AAV8 lime, AAV9 coral, AAV6 tan&#39;</span>
</pre></div>
</div>
<p>The following output will be generated, while the newly created html file is automatically displayed (thanks to option -d) in the default web browser:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Colorized</span> <span class="n">version</span> <span class="n">of</span> <span class="n">parental</span> <span class="nb">map</span> <span class="n">file</span> <span class="n">saved</span> <span class="n">into</span> <span class="n">file</span><span class="p">:</span> <span class="n">test</span><span class="o">-</span><span class="n">par</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
<div class="section" id="arguments">
<h2>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h2>
<p>The same parameters are available in both modes (graphic and command-line) but some default values might differ.</p>
<ul class="simple">
<li><p>Version: -v</p></li>
</ul>
<blockquote>
<div><p>Whenever the argument -v is used, parent-map will only display the program version and exit, no matter whether other arguments are present or not. In graphic mode, version information is under ‘About’ in the main menu.</p>
</div></blockquote>
<ul class="simple">
<li><p>Help: -h</p></li>
</ul>
<blockquote>
<div><p>Whenever the -h argument is used, parent-map will only display the help message and exit, no matter whether other arguments are present or not (except -v). In graphic mode, the ‘Help’ menu item links to the documentation web page.</p>
</div></blockquote>
<ul class="simple">
<li><p>Variants: file name or path to file name</p></li>
</ul>
<blockquote>
<div><p>The variant file must be in FASTA format, no matter how many sequences are present. In graphic mode, a Browse button allows to select the file from any directory.</p>
</div></blockquote>
<ul class="simple" id="parents">
<li><p>Parents: file name or path to file name</p></li>
</ul>
<blockquote>
<div><p>The parent file must be in FASTA format unless a single parent is present, in which case raw sequence (with no name) is fine. Again, a Browse button is there in graphic mode. It is possible to change the position numbering by adding a number after the sequence name (separated by a blank space). That number will then become the position number of the first character in the sequence. If no umber is provided, the first character will be considered to be in position 1. This is useful when standard numberings need to be used but only a region of the gene is sequenced.</p>
</div></blockquote>
<ul class="simple">
<li><p>Output: -o</p></li>
</ul>
<blockquote>
<div><p>The output files will be saved in the same directory as the variant file. If this argument is used, the prefix entered will be the part of the file names before the -stats.tx, -par.txt, -def.txt and -aln.txt endings. If this argument is not used, a prefix will be automatically generated from the variants and parents file names, as well as some arguments and their values.</p>
</div></blockquote>
<ul class="simple">
<li><p>MinFragLen: -m</p></li>
</ul>
<blockquote>
<div><p>This is the minimum fragment length used to map variant regions against parental sequences. Default values are 6 for protein sequences and 18 for DNA sequences (sequences with non [A, T, G, C] characters are automatically defined as ‘protein’). It is recommended to keep the default values.</p>
</div></blockquote>
<ul class="simple">
<li><p>MinOverlap: -V</p></li>
</ul>
<blockquote>
<div><p>This is the minimum overlap length in the case of sequences produced by shuffling parental sequences. Default values are 2 for protein and 6 for DNA. It can be useful to decrease the value in cases where variants from multiple parents are not produced by overlapping randomly digested fragments, but rather by rationally designed combinations.</p>
</div></blockquote>
<ul class="simple">
<li><p>MaxNameSize: -s</p></li>
</ul>
<blockquote>
<div><p>Maximum length of sequence names before the name is replaced with a shorter generic name (‘S’ followed by a number corresponding to the sequence position in the FASTA file). Default value is 12.</p>
</div></blockquote>
<ul class="simple">
<li><p>SeqChars: -c</p></li>
</ul>
<blockquote>
<div><p>Number of sequence characters per line in alignments (-par.txt and -aln.txt files). Default value is 120.</p>
</div></blockquote>
<ul class="simple">
<li><p>LowerCase: -l</p></li>
</ul>
<blockquote>
<div><p>By default, all sequences are displayed in upper case. Use the -l argument to display sequences in lower case instead.</p>
</div></blockquote>
<ul class="simple">
<li><p>VRSides: -e</p></li>
</ul>
<blockquote>
<div><p>Number of characters to be included each side of variable regions in the -aln.txt file. Note that increasing the value results in neighboring regions merging. Modifying this value only alters the esthetics of the displayed alignments. Default value is 1.</p>
</div></blockquote>
<ul class="simple">
<li><p>Overwrite: -f</p></li>
</ul>
<blockquote>
<div><p>By default, if an output file with the same name already exists, parent-map exits with a warning to either change the output prefix or to use the -f argument. The -f argument forces overwriting of existing files with identical names without warning.</p>
</div></blockquote>
<ul class="simple">
<li><p>Symbols: -S</p></li>
</ul>
<blockquote>
<div><p>Symbols for identity (default: ‘.’) and gap (default: ‘-‘) in -par.txt and -aln.txt files. Symbols should be exactly 2 non identical, non alphabetical and non blank characters.</p>
</div></blockquote>
<ul class="simple">
<li><p>DisplayResults: -d</p></li>
</ul>
<blockquote>
<div><p>The output files are automatically displayed in a web browser by default in graphic mode only. Use the -d argument to automatically display in command-line mode. The option might not work, depending on the user’s operating system and browser configuration. The txt files might be displayed in the default application for text files instead, if the browser is unresponsive, or might not display at all. In any case, the 5 output files are always saved in the same directory as the input files, and can be opened at any time using any appropriate displaying application (a web browser is recommended). Note that if a word processor is used, a monospaced font (such as Courier New for example) needs to be applied to the contents for proper display. Note that even when using a monospaced font, Microsoft Word uses shorter blank spaces, which distorts the display. In this case, replace all blank spaces in the document with a monospaced blank space created by pressing the space bar while simultaneously holding both Shift and Ctrl keys.</p>
</div></blockquote>
<ul class="simple" id="colors">
<li><p>Colors: -p</p></li>
</ul>
<blockquote>
<div><p>When parent-map is used to analyze sequences (using the 2 input files), the -p (palette) argument is unnecessary (default colors are used) unless the user wishes to apply a specific color scheme, in which case -p must be followed by a list of parent color pairs in the following format (pairs separated by commas, blank space between parent and color, list surrounded by single or double quotes):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">p</span> <span class="s1">&#39;parent1 color1, parent2 color2, parent2 color3,...&#39;</span>
</pre></div>
</div>
<p>Parent names must be exactly as they appear in the -par.txt file, including the case.
Color names are case-insensitive and must be standard html colors, such as shown on this page: <a class="reference external" href="http://www.html-color-names.com/color-chart.php">http://www.html-color-names.com/color-chart.php</a></p>
<p>When sequence analysis is bypassed and parent-map is only used to colorize an existing -par.txt file (using the -C argument followed by the file name), the -p argument is required.</p>
</div></blockquote>
<ul class="simple">
<li><p>ColorizeOnly: -C</p></li>
</ul>
<blockquote>
<div><p>This option is to bypass the sequence analysis and to only colorize an existing -par.txt file. The file name (or path) needs to be entered after -C.</p>
</div></blockquote>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="output-files">
<h1>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h1>
<div class="section" id="statistics">
<h2>Statistics<a class="headerlink" href="#statistics" title="Permalink to this headline">¶</a></h2>
<p>The -stats.txt file shows a summary of the variant sequences main features, in table format. Sequences are sorted by number of parental sequences, then by main parent name, then by the number of insertion sites. Description of the table columns:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt><strong>Name</strong>:</dt><dd><p>Variant sequence name as appearing in the variant FASTA file, or as a shorter name made of ‘S’ followed by the sequence number in the original file in case the original name is longer than the default maximum name length.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Length</strong>:</dt><dd><p>Sequence lengths (number of amino acid residues for protein sequences, number of nucleotides for DNA sequences).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Parents</strong>:</dt><dd><p>Number of parental sequences detected as having contributed to the variant sequence.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Main</strong>:</dt><dd><p>Name of the main contributing parental sequence (the parent covering the largest area of the variant, in case of multiple parents).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Coverage</strong>:</dt><dd><p>% of the variant sequence covered by the main parent. Note that this is not the same as global similarity to the main parent sequence in the case of multiple parents. In that case, it’s about the parts of the variant sequence that are not already covered by other parents. It’s actually the % of positions identical to the main parent as displayed in the -par.txt file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Matches</strong>:</dt><dd><p>Number of sequence positions matched (as opposed to a % as in the previous column).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>ID%</strong>:</dt><dd><p>Global similarity of the variant sequence compared to the main parent sequence (% of identical positions). In the case of a single parent, this is identical to coverage (6th column). However, in the case of multiple parents, this is the % identity with the parent along the whole variant sequence (and not just the part covered by that parent and not by other parents as in column 6).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Identities</strong>:</dt><dd><p>Global similarity as the number of identical variant sequence positions compared with the main parent.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Ins_sites</strong>:</dt><dd><p>Number of sites in the variant sequence that have an insertion compared with the main parent.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Ins</strong>:</dt><dd><p>Total number of nucleotide or amino acid insertions in the whole variant sequence.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Del_sites</strong>:</dt><dd><p>Number of sites in the variant sequence that have a deletion compared with the main parent.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Dels</strong>:</dt><dd><p>Total number of nucleotide or amino acid deletions in the whole variant sequence.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Subs</strong>:</dt><dd><p>Total number of substitutions in the variant sequence compared with the main parent.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Other</strong>:</dt><dd><p>Number of sequence positions that could not be matched to any parent and could not be considered as a variation of one particular parent.</p>
</dd>
</dl>
</li>
</ol>
</div>
<div class="section" id="parental-mapping">
<h2>Parental mapping<a class="headerlink" href="#parental-mapping" title="Permalink to this headline">¶</a></h2>
<p>In the -par.txt file, each variant complete sequence is displayed along its contributing parental fragments. The parental mapping is simplified as much as possible, so that the variant is completely described using as few parents as possible and as few fragments as possible. Only the necessary parents are shown. If a fragment can be mapped equally to more than one parent, the parent that has the highest overall coverage is chosen. Identities are displayed in the parent lines using an identity symbol (‘.’ by default). If a mismatch occurs between 2 fragments mapped to the same parents or at a sequence end, it is considered a variation of the parent next to it and displayed as the parent sequence (substitution) or using the gap symbol (‘-‘ by default) in the parent line for insertions or in the variant line for deletions. The -par.txt file is most useful for variant sequences selected from shuffled libraries. Two files are generated: a black and white text file (name ending with -par.txt) and a colorized html version of the same file (name ending with -par.html). In the html file, each parent name and its corresponding identity fragments are colorized using a different color (whenever possible). The default color scheme uses 12 different colors. If the number of parents in the file is greater than 12, the same set of 12 colors is used on the 12 next parents and so on. If the user wishes to apply specific colors, the <a class="reference internal" href="#colors"><span class="std std-ref">-p argument</span></a> needs to be used followed by a list of parent color pairs, either at the time of the sequence analysis or later with the -C argument to specify the -par.txt file to be colorized again.</p>
</div>
<div class="section" id="sequence-definitions">
<h2>Sequence definitions<a class="headerlink" href="#sequence-definitions" title="Permalink to this headline">¶</a></h2>
<p>The -def.txt file shows comprehensive definitions of variant sequences in terms of their parent(s), using as few instructions as possible. This file is most useful for designing synthesis, PCR and assembly strategies for the reconstruction of the variants from the parents. For each variant, a 5-column table is displayed:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt><strong>Variant region</strong>:</dt><dd><p>Each region is defined as a range (ex. 86-147), a number (ex. 457) or two consecutive numbers separated by a slash (ex. 233/234). In the first case, the numbers represent sequence positions, and both are included in the region (ex. the 1-10 region includes all 10 first positions in the sequence). In the second case, the region is a single position. It the third case, the region is the empty interval between two consecutive positions, and shows the presence of a deletion.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Parent/feature</strong>:</dt><dd><p>In the case of sequence identity with a parent, the parent name is displayed. Otherwise, the name of the feature (substitution, insertion, deletion) is displayed. Unmatched regions that can not be assigned to a parent are shown as “unmatched”.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Parent region</strong>:</dt><dd><p>The corresponding region in the parent sequence. For substitution, insertions and deletions, the parent is the one surrounding the feature (or next to the feature if the feature is at a sequence end). Position numbering is the one defined in the parental sequence file as explained in <span class="xref std std-ref">parents</span> (first position in the sequence becomes the number next to the sequence name if a number if present).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Variant sequence</strong>:</dt><dd><p>When a substitution, insertion or unmatched region is present, the region sequence is displayed here.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Parent sequence</strong>:</dt><dd><p>In the case of substitutions or deletions, the corresponding parental sequence is displayed here.</p>
</dd>
</dl>
</li>
</ol>
</div>
<div class="section" id="alignments">
<h2>Alignments<a class="headerlink" href="#alignments" title="Permalink to this headline">¶</a></h2>
<p>The -aln.txt file consists of two parts: the first concerns sequences derived from a single parent, while the second part is about sequences derived from multiple parents. In each part, variants derived from the same main parents are aligned together against their main parent. Only regions that differ from the parent in any of the variants are shown, and the region position is indicated by a number (numbering is the one defined in the parental sequence file as explained in <span class="xref std std-ref">parents</span>). Sequence identity is shown using the identity symbol (default: ‘.’), gaps are shown using the gap symbol (default: ‘-‘). Insertions are indicated by a gap in the parent sequence, deletions by a gap in the variant sequence. For substitutions, the actual sequence is shown. Increasing the region area (using the -e argument) allows to show more sequence context and to combine neighboring regions. The -aln.txt file is most useful for variants derived from a single parent (such as from insertion or targeted substitution libraries), as it allows to compare variants in discrete regions, which can be much larger in number and size when variants are derived from multiple parents (as in shuffled libraries).</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>In order to explore the functionalities of parent-map, a set of protein sequences of evolved or designed AAV capsid variants described in peer-reviewed articles was assembled, as well as a set of the protein sequences of all wild-type parental AAV capsids from which they are derived. They can be downloaded by clicking on the following links:</p>
<blockquote>
<div><p><a class="reference external" href="https://sourceforge.net/projects/parent-map/files/example_variants.fasta/download">Variants</a></p>
<p><a class="reference external" href="https://sourceforge.net/projects/parent-map/files/example_parents.fasta/download">Parents</a></p>
</div></blockquote>
<p><a class="reference external" href="https://doi.org/10.1016/j.ymthe.2017.09.006">Variants SCH2 and SCH9</a> are derived from a chimeric library combining fragments from AAV2, AAV4, AAV5, AAV6, AAV8 and AAV9. Using default values, parent-map correctly identifies parental contributions from AAV2, AAV6, AAV8 and AAV9 for both variants.</p>
<p><a class="reference external" href="https://doi.org/10.1016/j.omtm.2019.10.007">Variant AAV-F</a> derives from an AAV9 peptide insertion library. Parent-map correctly identifies AAV9 as the single parent and detects peptide sequence FVVGQSY insertion between positions 588 and 589.</p>
<p><a class="reference external" href="https://dx.doi.org/10.1128%2FJVI.00254-08">Variant AAV-DJ</a> derives from a shuffled DNA library made with parental serotypes AAV2, AAV4, AAV5, AAV8 and AAV9. Parent-map correctly identifies parental contributions from AAV2, AAV8 and AAV9, and detects an A to T substitution at position 457 (corresponding to position 458 in parent AAV8).</p>
<p><a class="reference external" href="https://dx.doi.org/10.1038%2Fnbt.3440">Variant AAV-PHP.B</a> derives from an AAV9 peptide insertion library. Parent-map correctly identifies AAV9 as the single parent and detects peptide sequence TLAVPFK insertion between positions 588 and 589.</p>
<p><a class="reference external" href="https://dx.doi.org/10.1073%2Fpnas.0813365106">Variant AAV2.5T</a> derives from a capsid library generated by DNA shuffling of parents AAV2 and AAV5 and error-prone PCR. Parent-map correctly identifies parental contributions from AAV2 (VP1-specific N-terminal region) and AAV5 (remainder of the sequence) and an A to T substitution at position 582 (581 in parent AAV5).</p>
<p><a class="reference external" href="https://doi.org/10.1126/scitranslmed.3005708">Variant 7m8</a> derives from an AAV2 peptide insertion library. Parent-map correctly identifies AAV2 as the single parent and detects peptide sequence LALGETTRPA insertion between positions 587 and 588 (AAV2 numbering), as well as substitutions I to T at position 240 and V to I at position 718 (708 in AAV2).</p>
<p><a class="reference external" href="https://dx.doi.org/10.1038%2Fmt.2014.139">Variants LI-A and LI-C</a> derive from a synthetic library based on AAV2 with mutations at targeted positions. Parent-map correctly identifies AAV2 as the single parent, as well as the 14 and 4 substitutions in LI-A and LI-C respectively.</p>
<p><a class="reference external" href="https://doi.org/10.1016/j.ymthe.2017.09.021">Variant NP84</a> derives from a shuffled DNA library. Parent-map describes it as combining regions of AAV2, AAV3B and AAV6, as well as having substitutions K to E at 532 and R to G at 585 in the C-terminal AAV2-homolog region.</p>
<p><a class="reference external" href="https://dx.doi.org/10.1038%2Fgt.2016.62">Variant OLIG001</a> derives from a shuffled DNA and error-prone PCR library. Parent-map identifies parental contributions from AAV2, AAV6, AV8 and AAV9, as well as a E to K substitution at position 532 (position 533 in relevant parent AAV8) and an unmatched H at position 726 between regions homologous to AAV8 and AAV6.</p>
<p><a class="reference external" href="https://dx.doi.org/10.1038%2Fnbt.1599">Variant AAV2i8</a> is a rationally designed chimeric capsid based on AAV2 in which the receptor footprint region was replaced with corresponding region from AAV8. Parent-map successfully identifies AAV2 and AAV8 parental contributions.</p>
<p><a class="reference external" href="https://dx.doi.org/10.1016%2Fj.neuron.2016.09.021">Variant rAAV2-retro</a> derives from an AAV2 peptide insertion and error-prone PCR library. Parent-map correctly identifies AAV2 as the parent, the peptide sequence LADQDYTKTA insertion between positions 587 and 588, as well as the 2 substitutions N to D at position 382 and V to I at position 718 (708 in AAV2).</p>
<p><a class="reference external" href="https://dx.doi.org/10.1038%2Fmt.2010.234">Variant AAV2-sept-Y-F</a> is a rationally designed multiple mutant of AAV2 combining seven Y to F mutations on its surface. Parent-map correctly identifies AAV2 as the single parent, as well as the 7 Y to F substitutions at positions 252, 272, 444, 500, 700, 704 and 730.</p>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/parent-map.png" alt="Logo"/>
    
    <h1 class="logo logo-name">Parent-map</h1>
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=damienmarsic&repo=Parent-map&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="#usage">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#arguments">Arguments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#output-files">Output files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#statistics">Statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parental-mapping">Parental mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-definitions">Sequence definitions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#alignments">Alignments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#examples">Examples</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Parent-map documentation</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Damien Marsic.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/pm-doc.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>